onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -122,11 +123,187 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
+    if (fence) {
+      fence->BeforeUsingAsOutput(p_op_kernel->Node().GetExecutionProviderType(), queue_id);
+    }
+  }
+
+  return Status::OK();
+}
+
+static Status PostSynchronizeNodeOutputs(OpKernelContextInternal& op_kernel_context, int queue_id) {
+  for (int output_index = 0; output_index < op_kernel_context.OutputCount(); ++output_index) {
+    Fence_t fence = op_kernel_context.OutputFence(output_index);
+    if (fence) {
+      fence->AfterUsedAsOutput(queue_id);
+    }
+  }
+
+  return Status::OK();
+}
+
 Status SequentialExecutor::Execute(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
                                    const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
-                                   std::vector<OrtValue>& fetches,
-                                   const std::unordered_map<size_t, CustomAllocator>& fetch_allocators,
+                                   std::vector<OrtValue>& fetches, const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
                                    const logging::Logger& logger) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  ExecutionFrame frame{feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state};
+  if (exec_plan_size > 0) {
+    ORT_RETURN_IF_ERROR(Execute(session_state, feeds, fetch_mlvalue_idxs, logger, frame, 0, exec_plan_size - 1));
+  }
+  ORT_RETURN_IF_ERROR(FetchOutputsFromExecutionFrame(frame, session_state.GetTransferIntermediateTensorOwnership(), fetches, logger));
+  return Status::OK();
+}
+
+#ifdef ENABLE_TRAINING
+Status SequentialExecutor::ExecutePartial(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
+                                          const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
+                                          std::vector<OrtValue>& fetches, const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                                          const logging::Logger& logger, int64_t run_id) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_last_index = exec_plan_vec.size() - 1;
+  std::shared_ptr<ExecutionFrame> frame = nullptr;
+  std::map<int64_t, std::pair<std::unique_ptr<onnxruntime::ExecutionFrame>, size_t>>::iterator it;
+  size_t program_counter = 0;
+  //REVIEW(codemzs): We can get rid of this const cast by basically making partial graph exec manager APIs as SessionState.
+  PartialGraphExecutionManager& partial_graph_runs_manager = const_cast<SessionState&>(session_state).GetPartialGraphExecutionManager();
```

Comment:
>const_cast [](start = 61, length = 10)

the const_cast here is indicating, maybe SessionState is not the best place to host PartialGraphExecutionManager. 

Is it better to have it in the executor ? i.e SequentialExecutor